### PR TITLE
checkpointing bcs

### PR DESCRIPTION
**Pull request by Ig-dolci**
# Description

This PR handles the issue with the incorrect checkpointing of Dirichlet boundary conditions. Currently, during tape recomputation, the user-defined boundary condition (``DirichletBC._original_arg``) is being updated. This leads to the [issue #4206](https://github.com/firedrakeproject/firedrake/issues/4206).
In that issue, James suggested an approach to checkpoint the Dirichlet boundary condition. This PR implements that suggestion and includes a test to verify the fix.
This change fixes [issue #4206](https://github.com/firedrakeproject/firedrake/issues/4206).

### CODE REVIEW

**Comment by connorjward** [comment]
To fix CI I think you need to rebase on top of current `master`.

File: firedrake/adjoint_utils/dirichletbc.py [hunk 1]
```diff
@@ -42,7 +42,11 @@ def _ad_create_checkpoint(self):
...
         return deps[0]
 
-    def _ad_restore_at_checkpoint(self, checkpoint):
-        if checkpoint is not None:
-            self.set_value(checkpoint.saved_output)
+    def _ad_restore_at_checkpoint(self, bv):
+        if bv is not None:
+            bc = type(self)(
+                self.function_space(), bv.saved_output,
+                self.sub_domain)
```
Comment (pbrubeck) [inline firedrake/adjoint_utils/dirichletbc.py:49]:
```suggestion
            bc = bc.reconstruct(bv.saved_output)
```

File: firedrake/adjoint_utils/dirichletbc.py [hunk 2]
```diff
@@ -42,7 +42,9 @@ def _ad_create_checkpoint(self):
 
         return deps[0]
 
-    def _ad_restore_at_checkpoint(self, checkpoint):
-        if checkpoint is not None:
-            self.set_value(checkpoint.saved_output)
+    def _ad_restore_at_checkpoint(self, bv):
+        if bv is not None:
+            bc = bc.reconstruct(g=bv.saved_output)
```
Comment (pbrubeck) [inline firedrake/adjoint_utils/dirichletbc.py:47]:
```suggestion
            bc = self.reconstruct(g=bv.saved_output)
```

**Review by dham — APPROVED** [review]

**Comment by angus-g** [comment]
It still seems like this *should* pass the tests. It would be nice to see it merged to help unblock us a bit, if that's possible?
